lntest/itest/signer_test.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/stretchr/testify/require"
+)
+
+// testDeriveSharedKey checks the ECDH performed by the endpoint
+// DeriveSharedKey. It creates an ephemeral private key, performing an ECDH with
+// the node's pubkey and a customized public key to check the validity of the
+// result.
+func testDeriveSharedKey(net *lntest.NetworkHarness, t *harnessTest) {
+	ctxb := context.Background()
+
+	// Create an ephemeral key, extracts its public key, and make a
+	// PrivKeyECDH using the ephemeral key.
+	ephemeralPriv, err := btcec.NewPrivateKey(btcec.S256())
+	require.NoError(t.t, err, "failed to create ephemeral key")
+
+	ephemeralPubBytes := ephemeralPriv.PubKey().SerializeCompressed()
+	privKeyECDH := &keychain.PrivKeyECDH{PrivKey: ephemeralPriv}
+
+	// assertECDHMatch checks the correctness of the ECDH between the
+	// ephemeral key and the given public key.
+	assertECDHMatch := func(pub *btcec.PublicKey,
+		req *signrpc.SharedKeyRequest) {
+
+		ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+		resp, err := net.Alice.SignerClient.DeriveSharedKey(ctxt, req)
+		require.NoError(t.t, err, "calling DeriveSharedKey failed")
+
+		sharedKey, _ := privKeyECDH.ECDH(pub)
+		if !bytes.Equal(resp.SharedKey, sharedKey[:]) {
+			require.Equal(
+				t.t, resp.SharedKey, sharedKey,
+				"failed to derive the expected key",
+			)
+		}
+	}
+
+	nodePub, err := btcec.ParsePubKey(net.Alice.PubKey[:], btcec.S256())
+	require.NoError(t.t, err, "failed to parse node pubkey")
+
+	customizedKeyFamily := int32(keychain.KeyFamilyMultiSig)
+	customizedIndex := int32(1)
+	customizedPub, err := deriveCustomizedKey(
+		ctxb, net.Alice, customizedKeyFamily, customizedIndex,
+	)
+	require.NoError(t.t, err, "failed to create customized pubkey")
+
+	// Test DeriveSharedKey with no optional arguments. It will result in
+	// performing an ECDH between the ephemeral key and the node's pubkey.
+	req := &signrpc.SharedKeyRequest{EphemeralPubkey: ephemeralPubBytes}
+	assertECDHMatch(nodePub, req)
+
+	// Test DeriveSharedKey with a KeyLoc which points to the node's pubkey.
+	req = &signrpc.SharedKeyRequest{
+		EphemeralPubkey: ephemeralPubBytes,
+		KeyLoc: &signrpc.KeyLocator{
+			KeyFamily: int32(keychain.KeyFamilyNodeKey),
+			KeyIndex:  0,
+		},
+	}
+	assertECDHMatch(nodePub, req)
+
+	// Test DeriveSharedKey with a KeyLoc being set in KeyDesc. The KeyLoc
+	// points to the node's pubkey.
+	req = &signrpc.SharedKeyRequest{
+		EphemeralPubkey: ephemeralPubBytes,
+		KeyDesc: &signrpc.KeyDescriptor{
+			KeyLoc: &signrpc.KeyLocator{
+				KeyFamily: int32(keychain.KeyFamilyNodeKey),
+				KeyIndex:  0,
+			},
+		},
+	}
+	assertECDHMatch(nodePub, req)
+
+	// Test DeriveSharedKey with RawKeyBytes set in KeyDesc. The RawKeyBytes
+	// is the node's pubkey bytes. Because the KeyLoc is nil, it will end up
+	// using the node's pubkey to perform the ECDH.
+	req = &signrpc.SharedKeyRequest{
+		EphemeralPubkey: ephemeralPubBytes,
+		KeyDesc: &signrpc.KeyDescriptor{
+			RawKeyBytes: net.Alice.PubKey[:],
```

Comment:
I think we don't want this case to be possible. Because you specify an explicit pubkey, you don't expect there to be some fall back or implicit behavior. If one sets the `key_desc`, it should be correct. So we should add an extra check in `signer_server.go`, for example:
```
	if in.KeyDesc != nil && in.KeyDesc.KeyLoc == nil {
		return nil, fmt.Errorf("when setting key_desc the field " +
			"key_desc.key_loc must also be set")
	}
```

This test case can then move to the bottom to make sure the error is returned.

Reply:
fixed. Now we will always expect the user to specify the key family in `key_desc.key_loc`